File: client/scripts/core/Interface.js
/**
 * s2s.Interface class
 * @author Eric Rozell
 */

(function(s2s,$) {
	
	/**
	 * Constructor for Interface instances
	 * @param data an interface object
	 */
	s2s.Interface = function(data) 
	{
		this.obj = data;
	}

	/**
	 * Constructs a label for the Interface
	 * @return a string label
	 */
	s2s.Interface.prototype.getLabel = function()
	{
		return this.obj["label"];
	}

	/**
	 * Gets an identifier for the Interface
	 * @return a string ID
	 */
	s2s.Interface.prototype.getId = function()
	{
		return this.obj.uri;
	}

	/**
	 * Gets the type of the Interface
	 * @return a string ID for the interface type
	 */
	s2s.Interface.prototype.getTypes = function()
	{
		return this.obj['types'];
	}

	/**
	 * Gets the description of the Interface
	 * @return a string description
	 */
	s2s.Interface.prototype.getDescription = function()
	{
		return this.obj.comment;
	}

	/**
	 * Gets the output of the Interface
	 * @return a string ID for the output of the interface
	 */
	s2s.Interface.prototype.getOutput = function()
	{
		return this.obj.output;
	}

	/**
	 * Gets the input supported by the Interface
	 * @return a string ID for the input supported by the interface (may be null)
	 */
	s2s.Interface.prototype.getInput = function()
	{
		return this.obj.input;
	}

	/**
	 * Gets the default limit for the Interface
	 * @return an integer value for the default limit of the interface (may be null)
	 */
	s2s.Interface.prototype.getDefaultLimit = function()
	{
		return this.obj.limit;
	}

	/**
	 * Gets the class of the Object
	 * @return a class
	 */
	s2s.Interface.prototype.getClass = function()
	{
		return s2s.Interface;
	}
	
})(edu.rpi.tw.sesf.s2s,jQuery);

File: server/src/main/java/edu/rpi/tw/sesf/s2s/Interface.java
package edu.rpi.tw.sesf.s2s;

import java.util.HashMap;
import java.util.Map;
import java.util.Vector;

import org.json.JSONObject;

public abstract class Interface implements InstanceData {
	
	private static final long serialVersionUID = 4511027545999806172L;

	private boolean _fromCache = false;

	public abstract String getURI();
	public abstract Vector<String> getTypes();
	public abstract String getOutput();
	public abstract String getLabel();
	public abstract String getComment();
	public abstract String getInput();
	//TODO: find use cases for these
	public abstract int getLimit();

	public void setFromCache() {
		_fromCache   = true;
	}
	
	public boolean isFromCache() {
		return _fromCache;
	}
	
	@SuppressWarnings({ "rawtypes", "unchecked" })
	public JSONObject getJSON()
	{
		Map map = new HashMap();
		map.put("uri", getURI());
		map.put("types", getTypes().toArray());
		map.put("output", getOutput());
		if (getLabel() != null) map.put("label", getLabel());
		if (getComment() != null) map.put("comment", getComment());
		if (getLimit() != -1) map.put("limit", getLimit());
		if (getInput() != null) map.put("input", getInput());
		if (this.getDataSource().getId() != null) map.put("datasource", this.getDataSource().getId());
		map.put("cached", isFromCache());
		JSONObject json = new JSONObject(map);
		return json;
	}
}
